File: Data Structures & Algorithms/subsets/submission-1.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        results = [] * 2**n

        def _subsets(nums, i, take, arr = None):
            if not arr:
                arr = []
            if take:
                arr.append(nums[i])
            if i == n - 1:
                results.append(arr.copy())
                return
            _subsets(nums, i+1,False, arr.copy())
            _subsets(nums, i+1,True, arr.copy())
        _subsets(nums, 0, False)
        _subsets(nums, 0, True)

        return results

File: Data Structures & Algorithms/subarray-sum-equals-k/submission-1.java
//Take the current index j
// if total + arr[j] = k, count += 1. Increment j
// if total + arr[j] < k them expand to right and check
// if total + arr[j] > k, shrink from left till runningSum <= k
class Solution {
    public int subarraySum(int[] nums, int k) {
        int count = 0;
        int runningSum = 0;
        Map<Integer, Integer> freq = new HashMap<>();
        freq.put(0, 1);
        for(int i = 0; i < nums.length; ++i){
            runningSum += nums[i];
            count += freq.getOrDefault(runningSum - k, 0);
            freq.put(runningSum, freq.getOrDefault(runningSum, 0) + 1);
        }

        return count;
    }
}